src/ngraph/runtime/gpu/gpu_emitter.cpp

```diff
@@ -772,92 +775,25 @@ cudnnSetOpTensorDescriptor(opTensorDesc,
-                                                           0.0);
-                        }
-                        // descriptors for Nd tensors
                         else
                         {
-                            std::vector<size_t> dimensions = input_shape;
-                            auto compute_strides = [](const std::vector<size_t>& dim) {
-                                std::vector<size_t> strides(dim.size(), 1);
-                                std::copy(dim.begin() + 1, dim.end(), strides.begin());
-                                for (int64_t i = dim.size() - 2; i >= 0; i--)
-                                {
-                                    strides[i] *= strides[i + 1];
-                                }
-                                return strides;
-                            };
-
-                            kernel::emit_cudnnTensorNdDescriptor(writer,
-                                                                 input_desc,
-                                                                 tensor_type,
-                                                                 dimensions.size(),
-                                                                 dimensions,
-                                                                 compute_strides(dimensions));
-
-                            // mark reduced axes of input tensor for output tensor descriptor
-                            for (auto const& idx_dim : reduction_axes)
-                            {
-                                dimensions[idx_dim] = 1;
-                            }
-                            kernel::emit_cudnnTensorNdDescriptor(writer,
-                                                                 output_desc,
-                                                                 tensor_type,
-                                                                 dimensions.size(),
-                                                                 dimensions,
-                                                                 compute_strides(dimensions));
-                            // emit sum reduce operation
-                            kernel::emit_cudnnReduceTensor(writer,
-                                                           args[0],
-                                                           out[0],
-                                                           "CUDNN_REDUCE_TENSOR_ADD",
-                                                           tensor_type,
-                                                           "CUDNN_NOT_PROPAGATE_NAN",
-                                                           input_desc,
-                                                           output_desc,
-                                                           1.0,
-                                                           0.0);
+                            auto& cudnn_emitter =
```

Comment:
This is good, some of my thought:
```
int instance_idx = OpFactory.getOp().build_instance(sth);
writer << "OpFactory,getOp(instance_idx).run(input, output);
```
Basically we can create an instance with specialized parameters, and the emitter just run it.

This can be shared for all Ops, whether use cudnn/cuda can be managed in the Ops. 
Let me know how you think about it.

Reply:
I actually have it setup to do this already with the CUDAEmitter for Pad, see this for example: 
https://github.com/NervanaSystems/ngraph/blob/42911c8f7dd7aa398c0a5707235ca75b8b76a7f8/src/ngraph/runtime/gpu/cuda_emitter.cpp#L35.

The emitted kernel is specific to the tensor input and the requested padding and this is how the emitted kernel looks for test GPU.max_pool_2d_1channel_1image_padded:

```
more gpu_codegen/cuda_kernel_pad_i1__1__5__5_pb1__0_pa1__2_pi_codegen.cu
extern "C" __global__ void cuda_pad_i1__1__5__5_pb1__0_pa1__2_pi(float* in, float* out)
{
    size_t tid = blockIdx.x * blockDim.x + threadIdx.x;
    if (tid < 25)
    {
        size_t idx = 7 + tid % 5 + ((tid / 5) % 5) * 7 + ((tid / 25) % 5) * 49;
        out[idx] = in[tid];
    }
}
```

Notice how no variables are passed in other than input and output as you suggest, since they are all emitted into the writer for compilation. 

It is then called in the emitter via (https://github.com/NervanaSystems/ngraph/blob/42911c8f7dd7aa398c0a5707235ca75b8b76a7f8/src/ngraph/runtime/gpu/gpu_emitter.cpp#L819)

```
                        writer << "gpu::invoke_primitive(ctx, " << pad_index << ", ";
                        writer << "std::vector<void*>{" << args[0].get_name() << "}.data(), ";
                        writer << "std::vector<void*>{pad_buffer}.data()";
                        writer << ");\n";
                    }
```

The importance of the gpu::invoke_primitive function is that it is compiled in nGraph as `extern "C"` so that it provides a pure C bridge between the JIT runtime and the ngraph runtime. I believe this function is providing the interface that you suggest with the proposed ``OpFactory``.

Reply:
These code snippets are from PR #817 which depends on this PR.

Reply:
I noticed you already have the builder. I'm looking for the possibility to wrap all those into a more unified interface.

Since we already have ctx during the build, should we pass another one during the run?

My understanding is we just need to make the interface `extern "C"`, and it will work, correct me if I'm wrong.

Reply:
Ah, I _think_ I understand now, please do correct me if I am wrong:

Instead of 
```gpu_primitive_emitter::get_cuda_emitter()``` and ```gpu_primitive_emitter::get_cudnn_emitter()``` 

your suggestion is that we could possibly just have ```idx = gpu_primitive_emitter/OpFactory.getOp(idx).build(…)``` 

and for invocation instead of ```gpu::invoke_primitive(idx)``` 

we have ```writer<<OpFactory.getOp(idx).run(input,output)```.

This would be nice, to have the building and invocation using the same interface, however, I'm not sure how to do it safely. You are correct, the interface needs to be ```extern "C"```, but that precludes this kind of an interface,

```writer<<"OpFactory.getOp(idx).run(input,output)"```

because this is a C++ call. However, if what you are meaning is:

```
writer <<"invoke(cts,idx,input,output)";
...
extern "C" invoke(ctx,idx,input,output) 
{
   ctx->OpFactory.getOp(idx).run(input,output);
}
```

That would be doable, and is pretty close to the current implementation. The only reason to avoid doing so is that by having a C++ pointer exist in the runtime context, it seems to tell the user that it is safe to be called within the JIT code when it is actually not. This is why the std::vector<gpu::primitives*> is stored in the runtime context as a gpu::primitive**, so that the JIT code has no idea that it is actually coming from a C++ object. Only numbers are being passed. We could save it in the runtime context as a void* to avoid this problem, however. I see merits both ways.